small improvements

Ladicek opened the pull request:
- add `MethodTyped.parameterCount()`
- add `BlockCreator.newEmptyArray()` overloads that take an `int` size
- add `BlockCreator.throw_()` overloads that take an `Expr` message
- add `ConstructorDesc.of()` overloads that accept vararg parameter types
- validate the return type on creation of `ConstructorDesc`
- add `TestClassMaker.loadClass()`
- improve the exception message when an `Item` is used on an unexpected place

Ladicek commented:
Commits are best reviewed in isolation, although they are all pretty small. The last commit adds certain (nontrivial?) overhead, so I'd be willing to gate this behind a system property, but I found it tremendously helpful when figuring out what `Expr` exactly needs to be converted to a local variable.

dmlloyd commented on src/main/java/io/quarkus/gizmo2/creator/BlockCreator.java:
```diff
@@ -2539,7 +2571,7 @@ default void throw_(Class<? extends Throwable> type) {
     /**
      * Throw a new exception of the given type with a message.
      *
-     * @param type the exception type (must not be {@code null})
```
I think this edit hit the wrong line.

Ladicek replied:
Ouch, you're right, my bad. Will fix.

Ladicek replied:
Done.

dmlloyd commented on src/main/java/io/quarkus/gizmo2/creator/BlockCreator.java:
```diff
@@ -2550,6 +2582,20 @@ default void throw_(Class<? extends Throwable> type, String message) {
+    /**
+     * Throw a new exception of the given type with a message.
+     *
+     * @param type the exception type
```
Same here; `type` is not nullable but `message` is nullable.

Ladicek replied:
Done.

dmlloyd commented on src/main/java/io/quarkus/gizmo2/desc/ConstructorDesc.java:
```diff
@@ -37,7 +37,33 @@ static ConstructorDesc of(ClassDesc owner) {
+     * @return the constructor descriptor (not {@code null})
+     */
+    static ConstructorDesc of(ClassDesc owner, Class<?>... paramTypes) {
+        ClassDesc[] params = new ClassDesc[paramTypes.length];
+        for (int i = 0; i < paramTypes.length; i++) {
+            params[i] = Util.classDesc(paramTypes[i]);
+        }
```
Maybe `Stream.of(paramTypes).map(Util::classDesc).toArray(ClassDesc[]::new)`?

Ladicek replied:
Will do.

Ladicek replied:
Done.

dmlloyd commented on src/main/java/io/quarkus/gizmo2/impl/Item.java:
```diff
@@ -18,6 +18,7 @@
 import io.quarkus.gizmo2.impl.constant.ConstantImpl;
 
 public abstract non-sealed class Item implements Expr {
+    private final String creationSite = Util.callerOutsideGizmo();
```
I wonder if this might have a significant performance impact. We might want to disable it by default.

Ladicek replied:
Right, I was expecting this comment :-) I have no idea how I would quantify the impact, but the number of `Item`s we allocate is nontrivial, so the impact here is likely to be nontrivial as well. I'll add a system property.

Ladicek commented:
In addition to addressing the comments, I reworked the commit that updates `TestClassMaker`. The class has gained a method `forClass()` that returns a helper object of class `TestClassOps` with the same methods as `TestClassMaker`, except for the given class (and not for the last accepted class). Actually, the helper methods directly on `TestClassMaker` now delegate to `TestClassOps` created for the last accepted class. That feels more natural and more universal.
